File: solution/141/answer.py
class ListNode(object):
    def __init__(self, x):
        self.val = x
        self.next = None

    def printval(self):
        print self.val

class Solution(object):
    def hasCycle(self, head):
        fastindex = head
        slowindex = head
        while fastindex and slowindex:
            fastindex = fastindex.next
            slowindex = slowindex.next
            if not fastindex:
                return False
            fastindex = fastindex.next
            if fastindex is slowindex:
                return True
        return False

if __name__ == "__main__":
    a = ListNode(1)
    b = ListNode(2)
    c = ListNode(3)
    d = ListNode(4)
    a.next = b
    b.next = c
    c.next = d
    d.next = b
    s = Solution()
    result = s.hasCycle(a)
    print result

File: solution/141/answer.cpp
#include <stdio.h>
#include <set>

struct ListNode{
    int val;
    ListNode *next;
    ListNode(int x):val(x), next(NULL){}
};

class Solution{
public:
    bool hasCycle(ListNode *head){
        ListNode* Fasthead = head;
        ListNode* Slowhead = head;
        while(Fasthead && Slowhead){
            Fasthead = Fasthead->next;
            Slowhead = Slowhead->next;
            if(!Fasthead){return false;}
            Fasthead = Fasthead->next;
            if(Fasthead==Slowhead){
                return true;
            }
        }
        return false;
    }
};

int main(){
    ListNode a(1);
    ListNode b(2);
    ListNode c(3);
    ListNode d(4);
    ListNode e(5);
    a.next = &b;
    b.next = &c;
    c.next = &d;
    d.next = &e;
    e.next = &c;
    Solution s;
    bool x = s.hasCycle(&a);
    printf("%d\n", x);
}